master/pkg/ssh/ssh.go

```diff
@@ -21,34 +27,97 @@ type PrivateAndPublicKeys struct {
 }
 
 // GenerateKey returns a private and public SSH key.
-func GenerateKey(rsaKeySize int, passphrase *string) (PrivateAndPublicKeys, error) {
+func GenerateKey(conf config.SSHConfig) (PrivateAndPublicKeys, error) {
+	var generatedKeys PrivateAndPublicKeys
+	switch conf.KeyType {
+	case config.RSAKeyType:
+		return generateRSAKey(conf.RsaKeySize)
+	case config.ECDSAKeyType:
+		return generateECDSAKey()
+	case config.ED25519KeyType:
+		return generateED25519Key()
+	default:
+		return generatedKeys, errors.New("Invalid crypto system")
+	}
+}
+
+func generateRSAKey(rsaKeySize int) (PrivateAndPublicKeys, error) {
 	var generatedKeys PrivateAndPublicKeys
 	privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
 	if err != nil {
-		return generatedKeys, errors.Wrap(err, "unable to generate private key")
+		return generatedKeys, errors.Wrap(err, "unable to generate RSA private key")
 	}
 
 	if err = privateKey.Validate(); err != nil {
 		return generatedKeys, err
 	}
 
 	block := &pem.Block{
-		Type:  trialPEMBlockType,
+		Type:  rsaPEMBlockType,
 		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
 	}
 
-	if passphrase != nil {
-		// TODO: Replace usage of deprecated x509.EncryptPEMBlock.
-		block, err = x509.EncryptPEMBlock( //nolint: staticcheck
-			rand.Reader, block.Type, block.Bytes, []byte(*passphrase), x509.PEMCipherAES256)
-		if err != nil {
-			return generatedKeys, errors.Wrap(err, "unable to encrypt private key")
-		}
+	publicKey, err := sshlib.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return generatedKeys, errors.Wrap(err, "unable to generate RSA public key")
+	}
+
+	generatedKeys = PrivateAndPublicKeys{
+		PrivateKey: pem.EncodeToMemory(block),
+		PublicKey:  sshlib.MarshalAuthorizedKey(publicKey),
+	}
+
+	return generatedKeys, nil
+}
+
+func generateECDSAKey() (PrivateAndPublicKeys, error) {
+	var generatedKeys PrivateAndPublicKeys
+	// Curve size currently not configurable, using the NIST recommendation.
+	privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return generatedKeys, errors.Wrap(err, "unable to generate ECDSA private key")
+	}
+
+	privateKeyBytes, err := x509.MarshalECPrivateKey(privateKey)
+	if err != nil {
+		return generatedKeys, errors.Wrap(err, "unable to marshal ECDSA private key")
+	}
+
+	block := &pem.Block{
+		Type:  ecdsaPEMBlockType,
+		Bytes: privateKeyBytes,
 	}
 
 	publicKey, err := sshlib.NewPublicKey(&privateKey.PublicKey)
 	if err != nil {
-		return generatedKeys, errors.Wrap(err, "unable to generate public key")
+		return generatedKeys, errors.Wrap(err, "unable to generate ECDSA public key")
+	}
+
+	generatedKeys = PrivateAndPublicKeys{
+		PrivateKey: pem.EncodeToMemory(block),
+		PublicKey:  sshlib.MarshalAuthorizedKey(publicKey),
+	}
+
+	return generatedKeys, nil
+}
+
+func generateED25519Key() (PrivateAndPublicKeys, error) {
+	var generatedKeys PrivateAndPublicKeys
```

Comment:
same nit here

Reply:
The nit being that this could skip being declared at all? If so, I agree. I'd rather see

```go
return PrivateAndPublicKeys{}, errors.Wrap(err, "unable to generate ED25519 private key")
```

and then no `generatedKeys` ever declared.

Reply:
agree